lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,156 @@
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+
+    // In case provider name is not 'aws' and no other provider schema was introduced,
+    // then this means that no other provider plugin was loaded. In this case
+    // we need to throw an error.
+    if (
+      this.serverless.service.provider.name !== 'aws' &&
+      !this.schema.properties.provider.properties.name
+    ) {
+      this.handleErrorMessages([
+        `Unsupported provider name '${this.serverless.service.provider.name}'`,
+      ]);
+      return;
+    }
+
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors, userConfig);
+      this.handleErrorMessages(messages);
+    }
+  }
+
+  handleErrorMessages(messages) {
+    if (messages.length) {
+      const errorMessage = messages.join('\n     ');
+      if (this.serverless.service.configValidationMode === 'error') {
+        throw new this.serverless.classes.Error(`${ERROR_PREFIX} ${errorMessage}`);
+      } else {
+        for (const message of messages) {
+          this.serverless.cli.log(`${ERROR_PREFIX} ${message}`, 'Serverless', { color: 'red' });
+        }
+      }
+    }
+  }
+
+  defineTopLevelProperty(name, subSchema) {
+    this.schema.properties[name] = subSchema;
+  }
+
+  defineProvider(name, options = {}) {
+    const currentProvider = _.get(this.schema, 'properties.provider.properties.name.const');
+    if (
+      this.schema.properties.provider.properties.name &&
+      currentProvider &&
+      currentProvider !== name
+    ) {
+      logWarning(
+        [
+          `Validation schema for '${name}' provider is ignored because`,
+          `there is already a validation schema for '${currentProvider}' provider.`,
+        ].join(' ')
+      ); // crash in v2
+      return;
+    }
+
+    this.schema.properties.provider.properties.name = { const: name };
+
+    if (options.provider) {
+      addPropertiesToSchema(this.schema.properties.provider, options.provider);
+    }
+
+    if (options.function) {
+      addPropertiesToSchema(
+        this.schema.properties.functions.patternProperties[FUNCTION_NAME_PATTERN],
+        options.function
+      );
+    }
+
+    if (options.functionEvents) {
+      if (!Array.isArray(options.functionEvents)) throw new Error('functionEvents should be array');
+      for (const functionEvent of options.functionEvents) {
+        this.defineFunctionEvent(name, functionEvent.name, functionEvent.schema);
+      }
+    }
+  }
+
+  defineCustomProperties(configSchemaParts) {
+    addPropertiesToSchema(this.schema.properties.custom, configSchemaParts);
+  }
+
+  defineFunctionEvent(providerName, name, configSchema) {
+    const currentProvider = _.get(this.schema, 'properties.provider.properties.name.const');
+
+    if (currentProvider !== providerName) {
+      return;
+    }
```

Comment:
Idea I proposed was about validating against service config.

If given service is configured with different provider then we should silently ignore this extension.

And same way in `defineProvider`, allow to invoke it with different provider extensions (do not warn or crash as we initially planned), but accept only one for which given service is configured.

There can be also case where serverless is invoked not in service context, then we should just not extend schema with provider extensions

Reply:
> Idea I proposed was about validating against service config.

Sorry, I don't understand. What is wrong with the current implementation? (This does not include the fact that I forgot to update numerous event plugins with the new `defineFuctionEvent` API with prepending provider name. I'll fix that.)
 
> If given service is configured with different provider then we should silently ignore this extension.

What do you mean by saying that service is configured with different provider? What's your definition of a service? For me, it's like a name for a serverless.yml file. And if I correctly understand, the serverless.yaml file can have only one provider.

> And same way in `defineProvider`, allow to invoke it with different provider extensions (do not warn or crash as we initially planned), but accept only one for which given service is configured.

The code above achieves it, doesn't it? Or do you want me to remove log warning inside the `defineProvide` method? Why?

> There can be also case where serverless is invoked not in service context, then we should just not extend schema with provider extensions

Do you mean the Serveless Components type of `serverless.yaml`? I briefly looked at it, it appears there's no `service` or `provider` properties.  I thought we are on the final curve and the current implementation harmoniously fits into that.

Reply:
> What is wrong with the current implementation?

It doesn't validate against `provider.name` in `serverless.yml` but against given provider being defined in schema

> What's your definition of a service?

_service_ is a Serverless _project_, with configuration defined in `serverless.yml`

> Do you mean the Serveless Components type of serverless.yaml

No, I meant just `sls` being invoked in directory where there's no `serverless.yml`

Reply:
> It doesn't validate against provider.name in serverless.yml but against given provider being defined in schema

I see. Not sure it'll influence anyhthing, but I could define `currentProvider` constant with `this.serverless.service.provider.name`  instead of `_.get(this.schema, 'properties.provider.properties.name.const')`.

That'll fix that, right?

 > No, I meant just sls being invoked in directory where there's no serverless.yml

Thanks for pointing this out. Yes, in this case I believe validation should not be performed. So in `Service` class we could perform validation only if `userConfig.service` is not null.

Reply:
> That'll fix that, right?

I meant that we should extend schema only if `providerName === this.serverless.service.provider.name`, and silently abort if it's not the case.

Other question that came to me after your question is: How will config validation behave when user is using custom provider which doesn't come with any schema (which will be the case for all the custom providers at the start)

Ideally would should not show configuration errors (related to `provider` and other fields which we allow to extend) but prompt with a warning as:

```
You're relying on provider plugin which doesn't provide a schema for its config. Please report the issue at its bug tracker linking: [link to schema extensions documentation]
```

I guess we can implement it in following way:

Just before applying config validation, check if we have _provider_ definitions loaded (existence of `_.get(this.schema, 'properties.provider.properties.name.const')` will confirm). If not, then relax schema with `additionalProperties: true` on `provider` and other properties which can be extended, show above warning and follow with regular validation.

Reply:
Makes sense. This is actually the very case when the user enters `asw` in provider name. So I adjusted this case for this new context.

Implemented in https://github.com/serverless/serverless/pull/7335/commits/aeaa2d29cd51af3cfe7a79602f4c89c7232a8192